### ISSUE TITLE
impl Parser for tuples

### ISSUE THREAD

**Issue by epage** [Geal/nom#1417]
## Prerequisites

Here are a few things you should provide to help me understand the issue:

- Rust version : 1.55.0
- nom version : 7.0.0
- nom compilation features used: default

## Test case

We call the sequence function `tuple`, its a small step from there to go ahead an impl `Parser` for tuples.

I took this idea from [combine](https://docs.rs/combine/4.6.1/combine/trait.Parser.html) and found porting combine code to nom, this added a lot of extra noise, obscuring the actual grammar among all the nom syntax.

*This is forked from #1408*

An example parser
```rust
use nom::{
    branch, bytes::complete as bytes, character::complete as character, combinator, multi,
    sequence, AsChar, IResult as Outcome,
};

pub(crate) fn dec_int(input: &str) -> IResult<&str, &str> {
    combinator::recognize(sequence::tuple((
        combinator::opt(multi::alt((character::char('+'), character::char('-')))),
        multi::alt((
            character::char('0'),
            combinator::map(
                sequence::tuple((
                    character::satisfy(|c| ('1'..='9').contains(&c)),
                    multi::take_while(is_dec_digit_with_sep),
                )),
                |t| t.0,
            ),
        )),
    )))(input)
}
```
Ideal
```rust
use nom::{
    branch, bytes::complete as bytes, character::complete as character, combinator, multi,
    sequence, AsChar, IResult as Outcome,
};

pub(crate) fn dec_int(input: &str) -> IResult<&str, &str> {
    combinator::recognize((
        combinator::opt(multi::alt((character::char('+'), character::char('-')))),
        multi::alt((
            character::char('0'),
            combinator::map(
                (
                    character::satisfy(|c| ('1'..='9').contains(&c)),
                    multi::take_while(is_dec_digit_with_sep),
                ),
                |t| t.0,
            ),
        )),
    ))(input)
}
```

**Comment by Geal** [comment]
good idea!

**Comment by Geal** [comment]
fixed in #1425
